File: tick/optim/prox/src/prox_oscar.cpp
// License: BSD 3 clause

#include "prox_oscar.h"

ProxOscar::ProxOscar(double strength,
                     double ratio,
                     bool positive)
  : ProxSortedL1(strength, positive) {
  set_ratio(ratio);
}

ProxOscar::ProxOscar(double strength,
                     double ratio,
                     ulong start,
                     ulong end,
                     bool positive)
  : ProxSortedL1(strength, start, end, positive) {
  set_ratio(ratio);
}

const std::string ProxOscar::get_class_name() const {
  return "ProxOscar";
}

void ProxOscar::compute_weights(void) {
  if (!weights_ready) {
    ulong size = end - start;
    weights = ArrayDouble(size);
    for (ulong i = 0; i < size; i++) {
      weights[i] = strength * (ratio * (size - i - 1)  + 1);
    }
    weights_ready = true;
  }
}

double ProxOscar::get_ratio() const {
  return ratio;
}

void ProxOscar::set_ratio(double ratio) {
  if (ratio < 0) {
    TICK_ERROR("Ratio should be non-negative");
  } else {
    if (ratio != this->ratio) {
      weights_ready = false;
      this->ratio = ratio;
    }
  }
}

File: tick/optim/prox/prox_oscar.py
# License: BSD 3 clause

from tick.optim.prox.base import Prox
import numpy as np

from tick.optim.prox.build.prox import ProxOscar as _ProxOscar


class ProxOscar(Prox):
    """Proximal operator of the OSCAR penalization.
    This penalization combines L1 penalization with a clustering penalization,
    that induces exact equality of weights corresponding to correlated features,
    so that clusters can be represented by a single coefficient.
    This penalization is therefore particularly relevant in high-dimensional
    problems with strong features correlation.

    Parameters
    ----------
    strength : `float`
        Level of penalization

    ratio : `float`
        The Oscar ratio parameter, with ratio >= 0. For ratio = 0 this is L1
        regularization, while a large ratio provides only the clustering effect.

    range : `tuple` of two `int`, default=`None`
        Range on which the prox is applied. If `None` then the prox is
        applied on the whole vector

    positive : `bool`, default=`False`
        If True, apply an extra projection onto the set of vectors with
        non-negative entries

    Notes
    -----
    This penalization was introduced in
    * Simultaneous regression shrinkage, variable selection and clustering of
      predictors with OSCAR, by Bondell H.D. and Reich B.J., Biometrics. 2008

    It uses the stack-based algorithm for FastProxL1 from

    * SLOPE--Adaptive Variable Selection via Convex Optimization, by
      Bogdan, M. and Berg, E. van den and Sabatti, C. and Su, W. and Candes, E. J.
      arXiv preprint arXiv:1407.3824, 2014
    """

    _attrinfos = {
        "strength": {
            "writable": True,
            "cpp_setter": "set_strength"
        },
        "ratio": {
            "writable": True,
            "cpp_setter": "set_ratio"
        },
        "positive": {
            "writable": True,
            "cpp_setter": "set_positive"
        }
    }

    def __init__(self, strength: float, ratio: float, range: tuple=None,
                 positive: bool=False):
        Prox.__init__(self, range)
        self.strength = strength
        self.ratio = ratio
        self.positive = positive
        self.weights = None
        if range is None:
            self._prox = _ProxOscar(self.strength, self.positive)
        else:
            self._prox = _ProxOscar(self.strength, self.range[0], self.range[1],
                                    self.positive)

    def _call(self, coeffs: np.ndarray, t: float, out: np.ndarray):
        self._prox.call(coeffs, t, out)

    def value(self, coeffs: np.ndarray) -> float:
        """
        Returns the value of the penalization at ``coeffs``

        Parameters
        ----------
        coeffs : `numpy.ndarray`, shape=(n_coeffs,)
            The value of the penalization is computed at this point

        Returns
        -------
        output : `float`
            Value of the penalization at ``coeffs``
        """
        return self._prox.value(coeffs)
